src/Microsoft.Extensions.Hosting.Abstractions/BackgroundService.cs

```diff
@@ -0,0 +1,99 @@
+    {
+        private static Action<object> _cancelToken = CancelToken;
+
+        private Task _executingTask;
+        private readonly CancellationTokenSource _stoppingCts = new CancellationTokenSource();
+        private readonly CancellationTokenSource _cancelledCts = new CancellationTokenSource();
+
+        public BackgroundService()
+        {
+            CancelledToken = _cancelledCts.Token;
+        }
+
+        /// <summary>
+        /// This method is called when the <see cref="IHostedService"/> starts. The implementation should return a task that represents
+        /// the lifetime of the long running operation(s) being performed.
+        /// </summary>
+        /// <param name="stoppingToken">Triggered when <see cref="IHostedService.StopAsync(CancellationToken)"/> is called.</param>
+        /// <returns>A <see cref="Task"/> that represents the long running operations.</returns>
+        protected abstract Task ExecuteAsync(CancellationToken stoppingToken);
+
+        /// <summary>
+        /// Triggered if cancellation is triggered during <see cref="IHostedService.StartAsync(CancellationToken)"/> or <see cref="IHostedService.StopAsync(CancellationToken)"/>.
+        /// </summary>
+        public CancellationToken CancelledToken { get; }
+
+        /// <summary>
+        /// Triggered when the application host is ready to start the service.
+        /// </summary>
+        /// <param name="cancellationToken">Indicates that the start process has been aborted.</param>
+        public virtual Task StartAsync(CancellationToken cancellationToken)
+        {
+            using (cancellationToken.Register(_cancelToken, _cancelledCts))
+            {
+                // Store the task we're executing
+                _executingTask = ExecuteAsync(_stoppingCts.Token);
+
+                // If the task is completed then return it, this will bubble cancellation and failure to the caller
+                if (_executingTask.IsCompleted)
+                {
+                    return _executingTask;
+                }
+
+                // Otherwise it's running
+                return Task.CompletedTask;
+            }
+        }
+
+        /// <summary>
+        /// Triggered when the application host is performing a graceful shutdown.
+        /// </summary>
+        /// <param name="cancellationToken">Indicates that the shutdown process should no longer be graceful.</param>
+        public virtual async Task StopAsync(CancellationToken cancellationToken)
+        {
+            // Stop called without start
+            if (_executingTask == null)
+            {
+                return;
+            }
+
+            using (cancellationToken.Register(_cancelToken, _cancelledCts))
+            {
+                try
+                {
+                    // Signal cancellation to the executing method
+                    _stoppingCts.Cancel();
+                }
+                finally
```

Comment:
Why is this a finally block? If this throws why wait for _executingTask? If _executingTask throws then you loose the original exception.